texar/torch/data/tokenizers/sentencepiece_tokenizer.py

```diff
@@ -0,0 +1,289 @@
+    # pickled. We need to define __setstate__ here.
+    def __setstate__(self, d):
+        self.__dict__, self.vocab_file = d
+        self.sp_model = spm.SentencePieceProcessor()
+        self.sp_model.Load(self.vocab_file)
+
+    def save_vocab(self, save_dir: str) -> Tuple[str]:
+        r"""Save the sentencepiece vocabulary (copy original file) to
+        a directory.
+        """
+        if not os.path.isdir(save_dir):
+            raise ValueError("Vocabulary path ({}) should be a "
+                             "directory".format(save_dir))
+        out_vocab_file = os.path.join(
+            save_dir, self._VOCAB_FILE_NAMES['vocab_file'])
+
+        if os.path.abspath(self.vocab_file) != os.path.abspath(out_vocab_file):
+            copyfile(self.vocab_file, out_vocab_file)
+
+        return (out_vocab_file,)
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.sp_model)
+
+    def _map_text_to_token(self, text: str) -> List[str]:  # type: ignore
+        return self.sp_model.EncodeAsPieces(text)
+
+    def _map_token_to_id(self, token: str) -> int:
+        return self.sp_model.PieceToId(token)
+
+    def _map_id_to_token(self, index: int) -> str:
+        token = self.sp_model.IdToPiece(index)
+        return token
+
+    def map_token_to_text(self, tokens: List[str]) -> str:
+        r"""Maps a sequence of tokens (string) in a single string."""
+        out_string = self.sp_model.DecodePieces(tokens)
+        return out_string
+
+    @staticmethod
+    def default_hparams() -> Dict[str, Any]:
+        r"""Returns a dictionary of hyperparameters with default values.
+
+        * If `hparams['vocab_file']` is specified, the tokenizer is directly
+          loaded from the vocabulary file. In this case, all other
+          configurations in `hparams` are ignored.
+        * Otherwise, the tokenizer is trained based on `hparams['text_file']`.
+        * `hparams['vocab_file']` and `hparams['text_file']` can not be None
+          at the same time.
+
+        .. code-block:: python
+
+            {
+                "vocab_file": None,
+                "text_file": None,
+                "vocab_size": None,
+                "model_type": "unigram",
+                "bos_token": "<s>",
+                "eos_token": "</s>",
+                "unk_token": "<unk>",
+                "pad_token": "<pad>",
+            }
+
+        Here:
+
+        `"vocab_file"`: str or None
+            The path to a sentencepiece vocabulary file.
+
+        `"text_file"`: str or None
+            Comma separated list of input sentences.
+
+        `"vocab_size"`: int or None
+            Vocabulary size. Note that the vocabulary size is predetermined,
+            and it is used in the tokenizer training procedure.
+
+        `"model_type"`: str
+            Model algorithm to train the tokenizer. Available algorithms are:
+            ``unigram``, ``bpe``, ``word``, and ``char``.
```

Comment:
Add this to the class-level docstring, cuz it's one of the most key information that users should know.

Perhaps order in this way: ``bpe``, ``word``, and ``char``, and ``unigram``

Reply:
Add a short note saying that the ``bpe`` here is different from the original bpe encoding proposed here (https://github.com/rsennrich/subword-nmt)

(This is to make the docstring as self-contained as possible. Instead of asking users to refer to the sentencepiece docs)